src/tojson.jl

```diff
@@ -4,53 +4,120 @@ using JSON
-    for s in ts.results
-        status = "pass"
-        message = nothing
-        for t in s.results
-            if typeof(t) == Test.Fail
+    if length(ts.results) == 1 && ts.results[1] isa Test.Error
+        # There has been a syntax error or similar and no tests have run.
+        # Otherwise ts.results[1] will be a ReportingTestSet.
+        return JSON.json(Dict(
+            "version" => 2,
+            "status" => "error",
+            "message" => ts.results[1].backtrace,
+            "tests" => [],
+        ), 4)
+    end
+
+    any_failed = false
+
+    """
+        walk!(prefix, tests, testset)
+
+    Walk the tree of testsets, pushing Dicts to `tests` describing each test
+    result. Returns nothing.
+    """
+    function walk!(tests, prefix, testset)
+        name = isempty(prefix) ? testset.description : "$prefix » $(testset.description)"
+
+        # Should we number the tests?
+        number_tests = count(x -> x isa Test.Result, testset.results) > 1
+
+        for (n, result) in enumerate(testset.results)
+            status = nothing
+            message = nothing
+            test_code = nothing
+
+            if result isa Test.Pass
+                status = "pass"
+                message = nothing
+            elseif result isa Test.Fail
                 status = "fail"
-                message = string(t)
-            elseif typeof(t) == Test.Error
+                message = string(result)
+                test_code = "@test " * result.orig_expr
+            elseif result isa Test.Error
                 status = "error"
-                message = t.backtrace
-
-                # status "error" supersedes "fail", therefore the loop does not need to continue
-                break
+                message = result.backtrace
+                test_code = "@test " * result.orig_expr
+            elseif result isa Test.AbstractTestSet
+                # Descend into nested test sets
+                child_has_failed = walk!(tests, name, result)
+                continue
+            else
+                error("Unknown testset.results item: $result")
             end
-        end
 
-        # test object
-        push!(tests, Dict(
-            "name" => s.description,
-            "status" => status,
-            "message" => message,
-            "output" => isempty(output) ? nothing : first(output, 500), # only show the first 500 characters
-        ))
-    end
+            any_failed = any_failed || status in ("fail", "error")
 
-    status = "pass"
-    message = nothing
-    for t in tests
-        if t["status"] == "fail"
-            status = t["status"]
-        elseif t["status"] == "error"
-            status = t["status"]
-            message = t["message"]
-
-            # status "error" supersedes "fail", therefore the loop does not need to continue
-            break
+            push!(tests, Dict(filter( ((k, v),) -> !isnothing(v), (
+                "name" => number_tests ? "$name.$n" : name,
```

Comment:
This is a change from the previous behaviour: previously only one test per testset was reported, now every test is reported separately.